### ISSUE TITLE
Cancelation Problem

### ISSUE THREAD

**Issue by emil14** [nevalang/neva#813]
In Go there's a way to stop goroutine from executing. It's a simple pattern where all you need to do is to pass a channel to that function. That functions must guarantee (compiler doesn't enforce though) to stop as soon as channel will be closed. 

Go design extends this pattern with first class support for `for` loops where you can range over channel messages. For loop will automatically break after channel is closed. Another construct that is frequently used with this pattern is `select` with case for cancelation channel. Go does even next step and implements half of this pattern with `context.Context` data-type that has `Done()` method that returns a channel you can block on. As soon as context is closed (by timeout or manual call to `cancel` func), the channel that `Done()` returns also closes.

Go however solves two problems with this data-type, it also allows to use it as a context in react-sense i.e. to share data without explicit drilling. That is a separate problem #812 that doesn't have to be solved in Nevalang exactly this way. Even thought it could be.

**Comment by default-writer** [comment]
Indeed, that will simplify coding for loops but in terms of clean and nice graceful shutdown it might me required to customise  this process and gave programmer the ability to exit on his own conditions or his own cases making the default behaviour works only if custom cancellation token processing omitted completely

**Comment by default-writer** [comment]
Please take a look into the ##custodians section: https://nikhilism.com/post/2023/racket-beyond-languages/

**Comment by default-writer** [comment]
Also, in general about CML and different languages and ideas: https://wingolog.org/archives/2016/09/21/is-go-an-acceptable-cml

**Comment by emil14** [comment]
> Indeed, that will simplify coding for loops but in terms of clean and nice graceful shutdown it might me required to customise this process and gave programmer the ability to exit on his own conditions or his own cases making the default behaviour works only if custom cancellation token processing omitted completely in user space code and not in library code they are using

@default-writer thanks, this issue doesn't propose (yet) some solution e.g. automatic cancelation (I don't have clear vision how this could work at the moment), but your point is valid - in case there will be something like that, we need to be sure it scales well for cases where custom logic is needed
